stonesoup/updater/probability.py

```diff
@@ -0,0 +1,192 @@
+        weights = np.asarray(posterior_state_weights)
+
+        # Reduce mixture of states to one posterior estimate Gaussian.
+        post_mean, post_covar = gm_reduce_single(means, covars, weights)
+
+        return post_mean, post_covar
+
+    def _update_via_innovation(self, hypotheses, **kwargs):
+        """Of n hypotheses there should be 1 prediction (a missed detection hypothesis) and n-1
+        different measurement associations. The update proceeds as described above.
+
+        Parameters
+        ----------
+        hypotheses : :class:`~.MultipleHypothesis`
+            As in :meth:`update` method
+         **kwargs : various
+            These are passed to :meth:`predict_measurement`
+
+        Returns
+        -------
+        : :class:`~.StateVector`
+            The mean of the reduced/single Gaussian
+        : :class:`~.CovarianceMatrix`
+            The covariance of the reduced/single Gaussian
+        """
+
+        for n, hypothesis in enumerate(hypotheses):
+            # Check for the existence of an associated measurement. Because of the way the
+            # hypothesis is constructed, you can do this:
+            if not hypothesis:
+                hypothesis.measurement_prediction = self.predict_measurement(
+                    hypothesis.prediction, **kwargs)
+                innovation = hypothesis.measurement_prediction.state_vector - \
+                    hypothesis.measurement_prediction.state_vector  # is zero in this case
+                posterior_covariance, kalman_gain = self._posterior_covariance(hypothesis)
+                # Add the weighted prediction to the weighted posterior
+                posterior_covariance = float(hypothesis.probability) * \
+                    hypothesis.prediction.covar + (1 - float(hypothesis.probability)) * \
+                    posterior_covariance
+                posterior_mean = copy(hypothesis.prediction.state_vector)
+            else:
+                innovation = hypothesis.measurement.state_vector - \
+                             hypothesis.measurement_prediction.state_vector
+
+            # probably exists a less clunky way of doing this using exists() or overwritten +=
+            # All these floats should be redundant if/when the bug in Probability.__mult__() is
+            # fixed.
+            if n == 0:
+                sum_of_innovations = float(hypothesis.probability) * innovation
+                sum_of_weighted_cov = float(hypothesis.probability) * (innovation @ innovation.T)
+            else:
+                sum_of_innovations += float(hypothesis.probability) * innovation
+                sum_of_weighted_cov += float(hypothesis.probability) * (innovation @ innovation.T)
+
+        posterior_mean += kalman_gain @ sum_of_innovations
+        posterior_covariance += \
+            kalman_gain @ (sum_of_weighted_cov - sum_of_innovations @ sum_of_innovations.T) \
+            @ kalman_gain.T
```

Comment:
Do these lines work if line 162 is False? I can't work out where posterior_mean, posterior_covariance or kalman_gain would come from in that case.

Reply:
Ignore me, I think I've got it now - the first hypothesis always a MissedDetection hypothesis right?

Reply:
> the first hypothesis always a MissedDetection hypothesis right?
>
It seems to be but that's a matter for the `associator.associate()` method. I think this should work whatever the order; it exploits the fact that `not hypothesis` returns True for a missed detection (not sure why that is) and calculates the required mean and covariance, and attaches a 0-innovation.